File: src/splitter_ffi.erl
-module(splitter_ffi).
-export([new/1, split/2, split_before/2, split_after/2, would_split/2]).

-spec new(Patterns) -> R when
    Patterns :: [binary()],
    R :: binary:cp().
new([]) ->
    empty_splitter;
new(Patterns) ->
    binary:compile_pattern(Patterns).

-spec split(Splitter,String) -> R when
    Splitter :: empty_splitter | binary:cp(),
    String :: binary(),
    R :: {binary(),binary(),binary()}.
split(empty_splitter, String) ->
    {<<>>, <<>>, String};
split(Splitter, String) ->
    case binary:match(String, Splitter) of
        nomatch -> {String, <<"">>, <<"">>};  % No delimiter found
        {Index, Length} ->
            {binary:part(String, 0, Index),
             binary:part(String, Index, Length),
             binary:part(String, Index + Length, byte_size(String) - Index - Length)}
    end.

-spec split_before(Splitter,String) -> R when
    Splitter :: empty_splitter | binary:cp(),
    String :: binary(),
    R :: {binary(),binary()}.
split_before(empty_splitter, String) ->
    {<<>>, String};
split_before(Splitter, String) ->
    case binary:match(String, Splitter) of
        nomatch -> {String, <<"">>};  % No delimiter found
        {Index, _Length} ->
            {binary:part(String, 0, Index),
             binary:part(String, Index, byte_size(String) - Index)}
    end.

-spec split_after(Splitter,String) -> R when
    Splitter :: empty_splitter | binary:cp(),
    String :: binary(),
    R :: {binary(),binary()}.
split_after(empty_splitter, String) ->
    {<<>>, String};
split_after(Splitter, String) ->
    case binary:match(String, Splitter) of
        nomatch -> {String, <<"">>};  % No delimiter found
        {Index, Length} ->
            SplitPoint = Index + Length,
            {binary:part(String, 0, SplitPoint),
             binary:part(String, SplitPoint, byte_size(String) - SplitPoint)}
    end.

-spec would_split(Splitter,String) -> R when
    Splitter :: empty | binary:cp(),
    String :: binary(),
    R :: {binary(),binary()}.
would_split(empty_splitter, _String) ->
    false;
would_split(Splitter, String) ->
    binary:match(String, Splitter) =/= nomatch.

File: src/splitter_ffi.mjs
import {
  List$Empty,
  List$NonEmpty,
  List$isNonEmpty,
  List$NonEmpty$first,
  List$NonEmpty$rest,
} from "./gleam.mjs";

export function make(patterns) {
  let pattern = "";
  let cursor = patterns;
  while (List$isNonEmpty(cursor)) {
    if (pattern !== "") pattern += "|";
    pattern += escapeRegExp(List$NonEmpty$first(cursor));
    cursor = List$NonEmpty$rest(cursor);
  }
  return new RegExp(pattern);
}

export function split(splitter, string) {
  const match = string.match(splitter);

  if (!match) return [string, "", ""]; // No delimiter found

  const index = match.index;
  const delimiter = match[0];

  return [
    string.slice(0, index),
    delimiter,
    string.slice(index + delimiter.length),
  ];
}

export function split_before(splitter, string) {
  const match = string.match(splitter);

  if (!match) return [string, ""]; // No delimiter found

  const split_point = match.index;
  return [string.slice(0, split_point), string.slice(split_point)];
}

export function split_after(splitter, string) {
  const match = string.match(splitter);

  if (!match) return [string, ""]; // No delimiter found

  const split_point = match.index + match[0].length;
  return [string.slice(0, split_point), string.slice(split_point)];
}

export function would_split(splitter, string) {
  // According to https://tc39.es/ecma262/#sec-escaperegexppattern
  // RegExp.source must escape "" as "(?:)"
  // i.e. this checks for the empty splitter
  if (splitter.source === "(?:)") return false
  return splitter.test(string)
}

function escapeRegExp(string) {
  return string.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

export function split_all(splitter, string) {
  let list = List$Empty();
  if (splitter.source === "(?:)") return List$NonEmpty(string, list);
  const parts = string.split(splitter);
  while (parts.length) {
    list = List$NonEmpty(parts.pop(), list);
  }
  return list;
}
